File: server/utils/exec.go
package utils

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
	"syscall"
)

var UseSudo bool

func RunCommand(isElevated bool, command string, args ...string) (string, error) {
	if isElevated {
		args = append([]string{command}, args...)
		command = "sudo"
	}

	LogLn(fmt.Sprintf("Running: %s %s", command, strings.Join(args, " ")))

	cmd := exec.Command(command, args...)
	// cmd.Env = append(os.Environ(), "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin")
	cmd.Stderr = logFile
	cmd.SysProcAttr = &syscall.SysProcAttr{
		Setpgid: true, // Create a new process group
	}
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}

	return string(out), err
}

func StartCommand(isElevated bool, command string, args ...string) (*exec.Cmd, error) {
	if isElevated {
		args = append([]string{command}, args...)
		command = "sudo"
	}

	LogLn(fmt.Sprintf("Starting: %s %s", command, strings.Join(args, " ")))

	cmd := exec.Command(command, args...)
	cmd.Stdout = logFile
	cmd.Stderr = logFile
	cmd.SysProcAttr = &syscall.SysProcAttr{
		Setpgid: true, // Create a new process group
	}
	err := cmd.Start()
	if err != nil {
		return nil, err
	}
	return cmd, nil
}

func IsRunning(cmdObject *exec.Cmd) bool {
	return cmdObject != nil && cmdObject.Process != nil && (cmdObject.ProcessState == nil || !cmdObject.ProcessState.Exited())
}

func SignalCmd(cmd *exec.Cmd, signal os.Signal) {
	if IsRunning(cmd) {
		cmd.Process.Signal(signal)
		// openvpnCmd.Wait()
	}
}

func CreateUser(username string) {
	cmd := exec.Command("/usr/sbin/adduser", "-S", "-D", "-H", "-h", "/dev/null", "-G", username, username)
	cmd.CombinedOutput()
}

func SignalRunning(pidFile string, signal os.Signal) bool {
	if _, err := os.Stat(pidFile); err != nil {
		LogF("PID file %s not found\n", pidFile)
		return false
	}

	file, err := os.Open(pidFile)
	if err != nil {
		LogF("Error opening PID file %s\n", pidFile)
		return false
	}
	defer file.Close()
	var pid int
	_, err = fmt.Fscanf(file, "%d", &pid)
	if err != nil {
		LogF("Error reading PID from file %s\n", pidFile)
		return false
	}

	err = SignalProcess(pid, syscall.Signal(0))
	if err != nil {
		LogF("Process with PID %d is not running\n", pid)
	}

	LogF("Sending signal %s to process with PID %d\n", signal, pid)
	return SignalProcess(pid, signal) == nil
}

File: server/utils/exec.py
import subprocess
import os
import signal
import sys
from .log import log_ln, get_log_file

use_sudo = False

def run_command(is_elevated, command, *args):
    cmd_args = list(args)
    if is_elevated or use_sudo:
        cmd_args.insert(0, command)
        command = "sudo"
    
    log_ln(f"Running: {command} {' '.join(cmd_args)}")
    
    try:
        result = subprocess.run(
            [command] + cmd_args,
            stderr=get_log_file() if get_log_file() != sys.stdout else subprocess.STDOUT,
            stdout=subprocess.PIPE,
            preexec_fn=os.setsid,
            text=True
        )
        return result.stdout, None
    except Exception as e:
        return "", e

def run_command_logged(is_elevated, command, *args):
    cmd_args = list(args)
    if is_elevated or use_sudo:
        cmd_args.insert(0, command)
        command = "sudo"
    
    log_ln(f"Running: {command} {' '.join(cmd_args)}")
    
    try:
        subprocess.run(
            [command] + cmd_args,
            stdout=get_log_file(),
            stderr=get_log_file(),
            preexec_fn=os.setsid
        )
        return None
    except Exception as e:
        return e

def start_command(is_elevated, command, *args):
    cmd_args = list(args)
    if is_elevated or use_sudo:
        cmd_args.insert(0, command)
        command = "sudo"
    
    log_ln(f"Starting: {command} {' '.join(cmd_args)}")
    
    try:
        process = subprocess.Popen(
            [command] + cmd_args,
            stdout=get_log_file(),
            stderr=get_log_file(),
            preexec_fn=os.setsid
        )
        return process, None
    except Exception as e:
        return None, e

def is_running(process):
    return process is not None and process.poll() is None

def signal_cmd(process, sig):
    if is_running(process):
        try:
            process.send_signal(sig)
        except ProcessLookupError:
            pass

def signal_process(pid, sig):
    try:
        if use_sudo:
            run_command(True, "/bin/kill", f"-{sig}", str(pid))
        else:
            os.kill(pid, sig)
        return None
    except Exception as e:
        return e

def signal_running(pid_file, sig):
    if not os.path.exists(pid_file):
        log_ln(f"PID file {pid_file} not found")
        return False
    
    try:
        with open(pid_file, "r") as f:
            pid = int(f.read().strip())
    except Exception as e:
        log_ln(f"Error reading PID from file {pid_file}")
        return False
    
    err = signal_process(pid, 0)
    if err is not None:
        log_ln(f"Process with PID {pid} is not running")
        
    log_ln(f"Sending signal {sig} to process with PID {pid}")
    return signal_process(pid, sig) is None
